### ISSUE TITLE
AutoZI: training evaluation

### ISSUE THREAD

**Issue by rlebron-bioinfo** [YosefLab/scVI#593]
Dear Yosef Lab,

First of all, I would like to thank you for your nice job in developing scVI 😄. I have some questions about AutoZI, could you help me out?

Basically, I want to know how to evaluate AutoZIVAE/UnsupervisedTrainer training in order to choose the most suitable hyperparameters (number of epochs and learning rate) for my datasets. A plot of the errors in training and/or testing would also help.

I would also appreciate a simple explanation of how AutoZI works and how to interpret its results. 

Thank you very much in advance.

All the very best,
Ricardo Lebrón

**Comment by adamgayoso** [comment]
Hi Ricardo,

In the basic tutorial, we show how to plot the negative Evidence lower bound (ELBO) for both training and test sets. You may use the test set ELBO as a held-out metric, useful for choosing hyperparameters. 

That said, we also have an [autotune module](https://nbviewer.jupyter.org/github/YosefLab/scVI/blob/master/tests/notebooks/autotune_advanced_notebook.ipynb) that can help you automatically choose the hyperparameters.

**Comment by rlebron-bioinfo** [comment]
> That said, we also have an [autotune module](https://nbviewer.jupyter.org/github/YosefLab/scVI/blob/master/tests/notebooks/autotune_advanced_notebook.ipynb) that can help you automatically choose the hyperparameters.

The goal is to **maximize** ELBO, right?

**Comment by adamgayoso** [comment]
Indeed it is, but we usually minimize the negative ELBO.

**Comment by rlebron-bioinfo** [comment]
How the _auto_tune_scvi_model()_ function decides which models/hyperparameters to evaluate next? Are they randomly chosen?

**Comment by oscarclivio** [comment]
Hi Ricardo,

Thank you for your questions.

Thank you also @adamgayoso for these first explanations.

So AutoZI is very similar to scVI but the ZINB is replaced by a mixture of NB and ZINB. The ZI probability in the ZINB has a lower bound which is a hyperparameter defined by the user. This allows us to decide whether zero-inflation applies to each gene under scrutiny or not. For this we study the mixture weight on a Bayesian fashion.

The general steps to analyze zero-inflation with AutoZI are the following :
- Train AutoZI on the given dataset
- Retrieve the parameters of the Beta distributions of the mixture weights for all genes
- Find zero-inflated genes using a decision criterion


There is a notebook tutorial about AutoZI detailing these steps and giving more complete explanations and which you may find here :
https://github.com/YosefLab/scVI/blob/master/tests/notebooks/AutoZI_tutorial.ipynb

Here is the paper for reference :
https://www.biorxiv.org/content/10.1101/794875v3

Here is a sample of code to see the evolution of train and test ELBOs during training, similarly to the basic tutorial notebook @adamgayoso referred to :

```
from scvi.models import AutoZIVAE

# Dataset

gene_dataset = CortexDataset(save_path=save_path, total_genes=558)

# Training hyperparameters

n_epochs = 400 if n_epochs_all is None else n_epochs_all
lr = 1e-3
use_batches = False
use_cuda = True

# Model hyperparameters

alpha_prior = 0.5
beta_prior = 0.5
minimal_dropout = 0.01

# Define AutoZIVAE and UnsupervisedTrainer

autozivae = AutoZIVAE(gene_dataset.nb_genes, alpha_prior=alpha_prior , beta_prior=beta_prior , minimal_dropout=minimal_dropout)
trainer = UnsupervisedTrainer(
    autozivae,
    gene_dataset,
    train_size=0.90,
    use_cuda=use_cuda,
    frequency=5,
)

# Train model
trainer.train(n_epochs=n_epochs, lr=lr)

# Plot error evolution
elbo_train_set = trainer.history["elbo_train_set"]
elbo_test_set = trainer.history["elbo_test_set"]
x = np.linspace(0, 400, (len(elbo_train_set)))
plt.plot(x, elbo_train_set, label="train")
plt.plot(x, elbo_test_set, label="test")
plt.legend()
```



It would be great to combine this with autotune though, I am not sure this is implemented yet...

**Comment by adamgayoso** [comment]
I'm closing this due to inactivity, but please feel free to reopen.

**Comment by rlebron-bioinfo** [comment]
I've some questions:
1) How to save and load the models generated by _auto_tune_scvi_model()_?
2) How can I make a comparison of the tested models? (e.g. training error plots)
3) How can I apply the best model to test datasets?
